File: tfp-did-mobile-sdk-detectid-plugin-react-native/ios/didbridge/DIDModule.m
//
//  DIDModule.m
//  DIDReact
//
//  Created by Camilo Ortegon on 10/7/22.
//

#import <Foundation/Foundation.h>

#import "React/RCTBridgeModule.h"
#import "React/RCTEventEmitter.h"

@interface RCT_EXTERN_MODULE(DIDModule,NSObject)

RCT_EXTERN_METHOD(getDeviceID:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(getMaskedAppInstanceID:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(getMobileID:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(didRegistration:(NSString)url resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(didRegistrationByQRCode:(NSString)url resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(updateGlobalConfig:(NSString)accountJson)
RCT_EXTERN_METHOD(isValidPayload:(NSDictionary)payload callback:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(subscribePayload:(NSDictionary)payload)
RCT_EXTERN_METHOD(receivePushServiceId:(NSString)data)
RCT_EXTERN_METHOD(setApplicationName:(NSString)data)

@end

File: tfp-did-mobile-sdk-detectid-plugin-react-native/ios/didbridge/DIDModule.swift
//
//  DIDModule.swift
//  DIDReact
//
//  Created by Camilo Ortegon on 10/7/22.
//

import Foundation
import didm_sdk

@objc(DIDModule)
class DIDModule: NSObject {
  
  @objc
  static func requiresMainQueueSetup() ->Bool {
    return false;
  }
  
  @objc
  func getDeviceID(_ callback:@escaping RCTResponseSenderBlock) {
    callback([(DetectID.sdk() as? DetectID)?.getDeviceID() ?? ""])
  }
  
  @objc
  func getMaskedAppInstanceID(_ callback:@escaping RCTResponseSenderBlock) {
    callback([(DetectID.sdk() as? DetectID)?.getMaskedAppInstanceID() ?? ""])
  }
  
  @objc
  func getMobileID(_ callback:@escaping RCTResponseSenderBlock) {
    callback([(DetectID.sdk() as? DetectID)?.getMobileID() ?? ""])
  }
  
  @objc
  func didRegistration(_ url: NSString, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) {
    (DetectID.sdk() as? DetectID)?.didRegistration(withUrl: url as String, onSuccess: {
      resolve(nil)
    }, onFailure: { error in
      reject("\(error.code)", error.localizedDescription, error)
    })
  }
  
  @objc
  func didRegistrationByQRCode(_ qrCode: NSString, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) {
    (DetectID.sdk() as? DetectID)?.didRegistration(byQRCode: qrCode as String, fromUrl: "", onSuccess: {
      resolve(nil)
    }, onFailure: { error in
      reject("\(error.code)", error.localizedDescription, error)
    })
  }
  
  @objc
  func updateGlobalConfig(_ accountJson: String) {
    guard let account = AccountModule.getAccount(fromJson: accountJson as String) else { return }
    (DetectID.sdk() as? DetectID)?.updateGlobalConfig(account)
  }
  
  @objc
  func isValidPayload(_ payload: [AnyHashable : Any], callback:@escaping RCTResponseSenderBlock) {
    callback(["\((DetectID.sdk() as? DetectID)?.isValidPayload(payload) ?? false)"])
  }
  
  @objc
  func subscribePayload(_ payload: [AnyHashable : Any]) {
    (DetectID.sdk() as? DetectID)?.subscribePayload(payload)
  }
  
  @objc
  func receivePushServiceId(_ data: String) {
  }
  
  @objc
  func setApplicationName(_ data: String) {
    (DetectID.sdk() as? DetectID)?.setApplicationName(data)
  }
  
}
